### PR TITLE
Boilerplate/parsing for Tuple_type

### PR DESCRIPTION
**Pull request by cjb129**

### CODE REVIEW

File: banjo/ast-type.hpp [hunk 1]
```diff
@@ -294,6 +294,29 @@ struct Array_type : Type
...
+  
+  Type const& type() const { return *t; }
+  Type&       type()       { return *t; }
+  
+  Expr const& extent() const { return *ext; }
+  Expr&       extent()       { return *ext; }
+  
+  Type_list* ty;
+  Expr* ext;
+  //FIXME: for printer...probably should be used somewhere
```
Comment (asutton) [inline banjo/ast-type.hpp]: Tuples are a sequence of types. They don't need an extra type pointer.

File: banjo/ast-type.hpp [hunk 2]
```diff
@@ -294,6 +294,29 @@ struct Array_type : Type
...
+  Type_list&       type_list()       { return *ty; }
+  
+  Type const& type() const { return *t; }
+  Type&       type()       { return *t; }
+  
+  Expr const& extent() const { return *ext; }
+  Expr&       extent()       { return *ext; }
+  
+  Type_list* ty;
+  Expr* ext;
```
Comment (asutton) [inline banjo/ast-type.hpp]: Tuples don't have an extent.

File: banjo/parse-type.cpp [hunk 3]
```diff
@@ -171,6 +182,7 @@ Type&
 //      decltype-type
 //      function-type
 //      '( unary-type )'
+//      '{ Type_list }'
```
Comment (asutton) [inline banjo/parse-type.cpp]: Should be type-list

File: banjo/printer.cpp [hunk 4]
```diff
@@ -481,6 +482,17 @@ Printer::postfix_type(Array_type const& t)
 }
 
 
+// FIXME: This is most likely very wrong
+void
+Printer::postfix_type(Tuple_type const& t)
+{
+  postfix_type(t.type());
+  token(lbrace_tok);
+  expression(t.extent());
```
Comment (asutton) [inline banjo/printer.cpp]: You need to print a comma delimited list of types.

File: banjo/printer.cpp [hunk 5]
```diff
@@ -481,6 +482,17 @@ Printer::postfix_type(Array_type const& t)
 }
 
 
+// FIXME: This is most likely very wrong
+void
+Printer::postfix_type(Tuple_type const& t)
```
Comment (asutton) [inline banjo/printer.cpp]: tuple-type is not postfix type. It is a primary type.

File: banjo/parse-type.cpp [hunk 6]
```diff
@@ -146,16 +146,27 @@ Parser::postfix_type()
...
+ match(rbracket_tok);
+ return on_array_type(t, e);
+}
+
  
+Type&
+Parser::tuple_type()
+{
+ match(lbrace_tok);
+ //FIXME: Needs to be a list of types? Not Expr
```
Comment (asutton) [inline banjo/parse-type.cpp]: Yes. Fortunately, this grammar is shared with function-type, so you can just call type_list.

**Comment by asutton** [comment]
Address comments, and update the request.

**Comment by cjb129** [comment]
I made the changes. I'm not sure how to update the request.

**Comment by cjb129** [comment]
Type_list iterator doesn't hold a Type? How do you access the type of an
object in the list?

**Comment by cjb129** [comment]
Nevermind. I think I understand it now.

**Pull request merged**